Make HFEM lower-D mesh work with distributed mesh

lindsayad opened the pull request:
We achieve this by not initially allowing remote element removal
when doing HFEM because the HFEM mesh construction process currently
relies on the mesh being in a serialized state. But then eventually
we can remove all the remote elements.

Closes #17484

lindsayad committed:
Make HFEM lower-D mesh work with serial distributed mesh

Without the changes here I could "run" the test that's modified
in this commit and not get an error in non-dbg modes, but I would
trigger assertions in `find_neighbors` when running in dbg mode.
With the changes here, I no longer hit those assertions

Refs #17484

YaqiWang commented:
Care to fix the line in `FEProblemBase` pointed by @yjung-anl during our conversation on Slack? Also with this change we can use distributed mesh for HFEM?

lindsayad committed:
Allow HFEM to run with distributed mesh

We achieve this by not initially allowing remote element removal
when doing HFEM because the HFEM mesh construction process currently
relies on the mesh being in a serialized state. But then eventually
we can remove all the remote elements

Closes #17484

lindsayad committed:
Fix bug in passed points

We were previously doing an inverse map to determine reference points
but then mistakenly passing the physical points to Assembly for
reinit. We want to make sure to pass the reference points

lindsayad commented:
> Also with this change we can use distributed mesh for HFEM?

Your current logic in `buildLowerDMesh` relies on the incoming mesh being in a serial state; I haven't done anything to fix that. However, since your logic doesn't rely on the mesh being *replicated* I removed any logic that relied on `is_replicated`. Consequently, you can now run HFEM with `--distributed-mesh`, but we will not remove remote elements until right before the `init_problem` task.

lindsayad commented:
> Care to fix the line in `FEProblemBase` pointed by @yjung-anl during our conversation on Slack

Done!

YaqiWang commented:
> Consequently, you can now run HFEM with `--distributed-mesh`, but we will not remove remote elements until right before the `init_problem` task.

I thought we remove remote elements at setup_mesh_complete?

roystgnr commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -2141,7 +2141,7 @@ FEProblemBase::reinitElemNeighborAndLowerD(const Elem * elem, unsigned int side,
       FEInterface::inverse_map(
           lower_d_elem_neighbor->dim(), FEType(), lower_d_elem_neighbor, qps, reference_points);
-      reinitLowerDElem(lower_d_elem_neighbor, tid, &qps);
+      reinitLowerDElem(lower_d_elem_neighbor, tid, &reference_points);
```
Whoa.  How did nobody catch this bug before?

lindsayad replied:
I don't know. I suspect the coverage report is going to tell me this line isn't hit. I suspect that we currently only ever call this method with the element that has the lower-D element associated with it, e.g. for non-refined meshes we are calling with the lower ID higher dimensional element etc.

roystgnr commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -215,6 +215,10 @@ MooseMesh::MooseMesh(const InputParameters & parameters)
 
+  if (getParam<bool>("build_all_side_lowerd_mesh"))
+    // Do not initially allow removal of remote elements
+    allowRemoteElementRemoval(false);
```
Where are we then restoring `allowRemoteElementRemoval(true)`?

lindsayad replied:
Currently nowhere. But this method marks `_need_delete = true` if someone calls with `allowRemoteElementRemoval(false)`, and that flag is checked during the `delete_remote_elements_after_late_geometric_ghosting` MOOSE action task.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1345981/) on 267b3da wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23407/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1345995/) on 267b3da wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>32ec9a</th>
      <th colspan=4>#23407 267b3d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.72%</td>
      <td>84.72%</td>
      <td>-0.00%</td>
      <td>71.43%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>82912</td>
      <td>82911</td>
      <td>-1</td>
      <td>5</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14950</td>
      <td>14950</td>
      <td>-</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23407/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23407/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23407/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 71.43% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad committed:
Don't build lower-d mesh when recovering
